### PR TITLE
port compiletest to use JSON output

### PR DESCRIPTION
**Pull request by nikomatsakis**
This uncovered a lot of bugs in compiletest and also some shortcomings
of our existing JSON output. We had to add information to the JSON
output, such as suggested text and macro backtraces. We also had to fix
various bugs in the existing tests.

Joint work with @jonathandturner.

r? @alexcrichton

### CODE REVIEW

**Comment by nikomatsakis** [comment]
OK, I think `make check` is passing locally now. I encountered a bunch of niggling errors towards the end of the cycle.

File: src/compiletest/json.rs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        vec![]
+    }
+}
+
+fn push_expected_errors(expected_errors: &mut Vec<Error>,
+                        diagnostic: &Diagnostic,
+                        file_name: &str) {
+    // We only consider messages pertaining to the current file.
+    let matching_spans =
+        || diagnostic.spans.iter().filter(|span| span.file_name == file_name);
```
Comment (alexcrichton) [inline src/compiletest/json.rs]: These are some... interesting choices of indentation?

File: src/compiletest/json.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    if line.chars().next() == Some('{') {
+        match json::decode::<Diagnostic>(line) {
+            Ok(diagnostic) => {
+                let mut expected_errors = vec![];
+                push_expected_errors(&mut expected_errors, &diagnostic, file_name);
+                expected_errors
+            }
+            Err(error) => {
+                println!("failed to decode compiler output as json: `{}`", error);
+                panic!("failed to decode compiler output as json");
```
Comment (alexcrichton) [inline src/compiletest/json.rs]: Could this/println be folded together?

File: src/libsyntax/errors/json.rs [hunk 3]
```diff
@@ -173,37 +204,93 @@ impl<'a> Diagnostic<'a> {
...
+                                                                 None,
+                                                                 vec![].into_iter(),
+                                                                 je)
+                               });
+                         Box::new(DiagnosticSpanMacroExpansion {
+                             span: call_site,
+                             macro_decl_name: bt.macro_decl_name,
+                             def_site_span: def_site_span,
+                         })
+                     });
```
Comment (alexcrichton) [inline src/libsyntax/errors/json.rs]: Out of curiosity, was this rustfmt-produced? This looks... pretty unreadable to me :(

File: src/tools/tidy/src/main.rs [hunk 4]
```diff
@@ -19,6 +19,11 @@ use std::path::{PathBuf, Path};
 use std::env;
 
 macro_rules! t {
+    ($e:expr, $p:expr) => (match $e {
+        Ok(e) => e,
+        Err(e) => panic!("{} failed on {} with {}", stringify!($e), ($p).display(), e),
+    });
```
Comment (alexcrichton) [inline src/tools/tidy/src/main.rs]: While I agree that it's nice to have a path name in error messages, it was somewhat intentional that this doesn't diverge from `try!` (as this shows up elsewhere). It doesn't really matter too much though.

File: src/test/run-make/tools.mk [hunk 5]
```diff
@@ -1,3 +1,6 @@
+# Default to `/tmp` if no other tmp directory is specified.
+TMPDIR ?= "/tmp"
```
Comment (alexcrichton) [inline src/test/run-make/tools.mk]: Were you running tests manually? In theory should should always be set by the harness running these tests.

**Comment by alexcrichton** [comment]
Changes pretty much all look good to me, just a few minor nits here and there. I'm pretty scared about the style this is introducing but no need to block anything on that.

File: src/tools/tidy/src/main.rs (hunk 4 above)
Comment (nikomatsakis) [inline src/tools/tidy/src/main.rs]: @alexcrichton the path name does not always show up elsewhere, or at least I couldn't find it; I added the path name to debug some problems I was seeing around symbolic links. I can remove it, but it seems a shame to not know what file is encountering a problem.

File: src/test/run-make/tools.mk (hunk 5 above)
Comment (nikomatsakis) [inline src/test/run-make/tools.mk]: > Were you running tests manually? In theory should should always be set by the harness running these tests.

I was just running `make check`.

File: src/libsyntax/errors/json.rs (hunk 3 above)
Comment (nikomatsakis) [inline src/libsyntax/errors/json.rs]: > Out of curiosity, was this rustfmt-produced? This looks... pretty unreadable to me :(

Heh, no. It's just how I generally format things. I guess in this case you could write

```rust
let def_site_span = bt.def_site_span.map(|sp| {
    ...
});
```

Iis that what you mean?

File: src/compiletest/json.rs (hunk 1 above)
Comment (nikomatsakis) [inline src/compiletest/json.rs]: Would you prefer this?

```rust
let matching_spans = || {
    diagnostic.spans.iter().filter(|span| span.file_name == file_name)
};
```

**Comment by nikomatsakis** [comment]
@alexcrichton 

> I'm pretty scared about the style this is introducing but no need to block anything on that.

Are you referring to the indentation, or something else?

File: src/test/run-make/tools.mk (hunk 5 above)
Comment (alexcrichton) [inline src/test/run-make/tools.mk]: Hm that's kinda scary, I'd rather leave this out for now as it's the wrong definition on Windows and perhaps investigate later what's going on locally for you?

File: src/tools/tidy/src/main.rs (hunk 4 above)
Comment (alexcrichton) [inline src/tools/tidy/src/main.rs]: Right yeah this is fine to leave for now, eventually we could perhaps add some wrappers around calling these functions in bootstrap (which add the path as a payload to the error) and then change this later.

File: src/libsyntax/errors/json.rs (hunk 3 above)
Comment (alexcrichton) [inline src/libsyntax/errors/json.rs]: Yeah I personally prefer avoiding rightward-drift as much as possible, the extra alignment here just seems a little confusing. This just looked the same as https://github.com/rust-lang-nursery/rustfmt/issues/439, but not a blocker at all!

File: src/compiletest/json.rs (hunk 1 above)
Comment (alexcrichton) [inline src/compiletest/json.rs]: Personally, yeah, but I'm mostly just randomly commenting :)

**Comment by alexcrichton** [comment]
> Are you referring to the indentation, or something else?

Yeah mostly how things are aligned to encourage rightward drift and how I have no idea how to configure an editor to automatically generate that style. That being said it's pretty minor.

Oh and meant to say, r=me, the only real change I'd like to see is `TMPDIR` removed as that's probably papering over something which should be fixed elsewhere, but otherwise seems fine.

**Comment by nikomatsakis** [comment]
@alexcrichton ok, I removed that TMPDIR change and I also tweaked the style a bit. :)

**Comment by nikomatsakis** [comment]
```
Linkcheck stage2 (i686-pc-windows-gnu)
panic_abort\fn.rust_eh_personality.html:50: broken link - src\panic_abort\lib.rs.html
panic_abort\fn.__rust_maybe_catch_panic.html:50: broken link - src\panic_abort\lib.rs.html
panic_abort\fn.__rust_start_panic.html:50: broken link - src\panic_abort\lib.rs.html
panic_abort\index.html:50: broken link - src\panic_abort\lib.rs.html
panic_unwind\fn.rust_eh_register_frames.html:50: broken link - src\panic_unwind\gcc\mod.rs.html
panic_unwind\fn.rust_eh_unregister_frames.html:50: broken link - src\panic_unwind\gcc\mod.rs.html
panic_unwind\fn.__rust_maybe_catch_panic.html:50: broken link - src\panic_unwind\lib.rs.html
panic_unwind\fn.__rust_start_panic.html:50: broken link - src\panic_unwind\lib.rs.html
panic_unwind\index.html:50: broken link - src\panic_unwind\lib.rs.html


command did not execute successfully: "C:\\bot\\slave\\auto-win-gnu-32-opt-rustbuild\\build\\obj\\build\\i686-pc-windows-gnu\\stage2-rustc\\i686-pc-windows-gnu\\release\\linkchecker.exe" "C:\\bot\\slave\\auto-win-gnu-32-opt-rustbuild\\build\\obj\\build\\i686-pc-windows-gnu\\doc"
expected success, got: exit code: 101


Makefile:40: recipe for target 'check' failed
Zombie process: "\Device\HarddiskVolume1\dojob.exe"
thread '<main>' panicked at 'found some broken links', C:\bot\slave\auto-win-gnu-32-opt-rustbuild\build\src\tools\linkchecker\main.rs:54
note: Run with `RUST_BACKTRACE=1` for a backtrace.
make: *** [check] Error 1
```

Hmm.

**Comment by nikomatsakis** [comment]
I think this rebase went awry

**Comment by nikomatsakis** [comment]
um. something is *definitely* messed up there...

File: src/tools/compiletest/src/json.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        vec![]
+    }
+}
+
+fn push_expected_errors(expected_errors: &mut Vec<Error>,
+                        diagnostic: &Diagnostic,
+                        file_name: &str) {
+    // We only consider messages pertaining to the current file.
+    let matching_spans = || {
+        diagnostic.spans.iter().filter(|span| span.file_name == file_name)
```
Comment (alexcrichton) [inline src/tools/compiletest/src/json.rs]: I think `file_name` has backslashes normalized away, but `span.file_name` does not, so on Windows this match never fires.

(discovered with @jonathandturner in person and he's working on a fix)

As a small note, using `Path` here may help as that equality takes path normalization (like `/` vs `\`) into account.

**Comment by alexcrichton** [comment]
@bors: r- 

oh wait this one was legit

**Comment by nikomatsakis** [comment]
Hallelujah! :)

**Pull request merged**